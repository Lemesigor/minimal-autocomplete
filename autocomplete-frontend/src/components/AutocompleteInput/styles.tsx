import { css } from '@emotion/react';

export const containerStyle = css`
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  padding-top: 5rem;
  `;

export const inputStyle = css`
    font-size: 1.125rem;
    color: #1f2937;
    border: 1px solid #1f2937;
    border-radius: 0.375rem;
    width: 15rem;
    padding: 0.25rem;
    margin-bottom: 0.5rem;
    transition: all 0.2s ease; 
  height: 30px;

    &:focus,
    &:hover {
        outline: none;
    }
`;