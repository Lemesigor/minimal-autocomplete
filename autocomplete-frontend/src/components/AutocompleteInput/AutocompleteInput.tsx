/** @jsxImportSource @emotion/react */
import {useEffect, useState} from "react";
import {containerStyle, inputStyle} from "./styles";
import {useRecentSearches} from "../../hooks/useRecentSearches";
import {useDebounce} from "../../hooks/useDebounce";
import {getAutocompleteData, Term} from "../../services/api/autocomplete";
import {AutocompleteList} from "../AutocompleteList/AutocompleteList";
import {retrieveFromStorage, saveToStorage, STORAGES} from "../../services/storage";


export const AutocompleteInput = () => {
    const [searchTerm, setSearchTerm] = useState<string>("");
    const [isInputFocused, setIsInputFocused] = useState<boolean>(false);
    const [suggestions, setSuggestions] = useState<Term[]>([]);
    const {recentSearches, addToRecentSearches} = useRecentSearches()
    const [isRecentSearchSuggestion, setIsRecentSearchSuggestion] = useState<boolean>(false)

    const debouncedSearchTerm = useDebounce({value: searchTerm, delay: 250})

    const shouldShowRecentSearchesList = !searchTerm && isRecentSearchSuggestion && recentSearches.length > 0

    const shouldShowSuggestionsList = isInputFocused && (shouldShowRecentSearchesList || suggestions.length)


    const handleSuggestionClick = (suggestion: Term) => {
        setSearchTerm(suggestion.value);
        addToRecentSearches({id: suggestion.id, value: suggestion.value});
    };

    useEffect(() => {
        const abortController = new AbortController();

        (async function getSuggestionsTerms() {
            if (!debouncedSearchTerm) {
                setSuggestions(recentSearches)
                setIsRecentSearchSuggestion(true)
                return
            }

            const cachedSearchTerms = retrieveFromStorage(debouncedSearchTerm, STORAGES.session)

            if (cachedSearchTerms) {
                setSuggestions(cachedSearchTerms as Term[])
                return
            }

            const signal = abortController.signal
            const termsListResult = await getAutocompleteData(debouncedSearchTerm, signal)
            setSuggestions(termsListResult ?? [])
            setIsRecentSearchSuggestion(false)

            saveToStorage(debouncedSearchTerm, termsListResult, STORAGES.session)
        })();
        return () => abortController.abort('Aborted due to new search term')
    }, [debouncedSearchTerm, recentSearches])

    return (
        <div css={containerStyle}>
            <input
                data-testid={"autocomplete-input"}
                css={inputStyle}
                placeholder="Search an English Word"
                type="text"
                value={searchTerm}
                onChange={(e) => setSearchTerm(e.target.value)}
                onFocus={() => setIsInputFocused(true)}
                onBlur={() => setIsInputFocused(false)}
            />
            {shouldShowSuggestionsList ?
                (<AutocompleteList items={suggestions}
                                   handleClick={handleSuggestionClick}
                                   isRecentSearchSuggestion={isRecentSearchSuggestion}/>): null}
        </div>
    );
};