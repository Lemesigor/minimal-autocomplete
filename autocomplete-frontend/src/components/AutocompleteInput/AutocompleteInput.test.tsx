import {render} from "@testing-library/react";

import {AutocompleteInput} from "./AutocompleteInput";

describe("AutocompleteInput", () => {
    it("should render the autocomplete input", () => {
        const {container} = render(<AutocompleteInput />);
        expect(container).toMatchSnapshot();
    });

    it("should have input field", () => {
        const {getByPlaceholderText} = render(<AutocompleteInput />);
        expect(getByPlaceholderText("Search an English Word")).toBeInTheDocument();
    })
});