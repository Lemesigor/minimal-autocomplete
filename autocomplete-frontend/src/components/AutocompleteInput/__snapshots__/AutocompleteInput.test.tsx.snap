// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteInput should render the autocomplete input 1`] = `
<div>
  <div
    class="css-wmd499"
  >
    <input
      class="css-y6v8d0"
      data-testid="autocomplete-input"
      placeholder="Search an English Word"
      type="text"
    />
  </div>
</div>
`;
